lua/plenary/async_lib2/async.lua

```diff
@@ -0,0 +1,110 @@
+local co = coroutine
+local vararg = require('plenary.vararg')
+local errors = require('plenary.errors')
+local traceback_error = errors.traceback_error
+local f = require('plenary.functional')
+
+local M = {}
+
+---because we can't store varargs
+local function callback_or_next(step, thread, callback, ...)
+  local stat = f.first(...)
+
+  if not stat then
+    error(string.format("The coroutine failed with this message: %s", f.second(...)))
+  end
+
+  if co.status(thread) == "dead" then
+    (callback or function() end)(select(2, ...))
```

Comment:
I think I'd prefer to just write this as:

```lua
if not callback then return end
callback(select(2, ...))
```

Don't think it matters a bunch but it just seems weird to create an empty function and call it again instead of just doing the same or condition above.